core/src/main/java/org/apache/calcite/rel/RelNode.java

```diff
@@ -407,6 +407,15 @@ RelNode copy(
    */
   void register(RelOptPlanner planner);
 
+  /**
+   * Indicates whether it is an enforcer operator, e.g. PhysicalSort,
+   * PhysicalHashDistribute, etc. As an enforcer, the operator must be
+   * created only when required traitSet is not satisfied by its input.
+   *
+   * @return Whether it is an enforcer operator
+   */
+  boolean isEnforcer();
```

Comment:
Does it make sense to add `default` implementation?

Reply:
Yes, will do.